python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,20 +1171,77 @@ class PyInferenceRequest
     ThrowIfError(TRITONSERVER_InferenceRequestSetDoubleParameter(
         triton_object_, key.c_str(), value));
   }
+
   void Cancel()
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    if (responses_.empty()) {
+      if (response_future_.get() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_.reset(new py::object(py_future));
```

Comment:
not sure I follow the logic here - why do we create a new object based on the passed in object? why not a reference on the original object?

second: how does the swap below work?

Reply:
I tried holding a raw pointer to the original object, but the pointer can no longer be dereferenced (segmentation fault) after the `GetNextResponse()` returned, so the only safe way of holding on to the future object after `GetNextResponse()` returns is to increment the reference count on the future object, which is achieved by "copying" it to a new object.

Reply:
Experiment: Increment the reference count on the py_future object, and see if the object can be remained valid after the function return.

Reply:
Can we have some kind of small memory leak/growth pytest test that checks memory before, does some inferences, and checks memory after?

Reply:
The implementation is updated to remove the pointer wrapper around `response_future_`, and more comments clarifying the flow among `GetNextResponse()` and `AddNextResponse()`: [Avoid extra malloc when tracking response_future_](https://github.com/triton-inference-server/core/pull/426/commits/c5ee14009b2c10434e3eefece50db40432da7314)
cc @nnshah1 @rmccorm4

Reply:
> Can we have some kind of small memory leak/growth pytest test that checks memory before, does some inferences, and checks memory after?

Do you think something like [this ](https://pytest-memray.readthedocs.io/en/latest/usage.html) makes sense? I think there will always be some small growth, but we can set the acceptable growth on the test case and have the test fail if it grows beyond what we set.

Reply:
Looks worth a try to me - I see it's used by quite a few big projects - https://github.com/search?q=pytest-memray&type=code

- pydantic
- datadog
- huggingface
- ...

I'm curious how much time overhead it adds to run.

Reply:
> but we can set the acceptable growth on the test case and have the test fail if it grows beyond what we set.

Let's make sure whatever threshold we set isn't tailored to the specific test case or number of iterations, and doesn't get hit if we increased the iterations on the test case or something